### PR TITLE
Merge k8s spike

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

Merge the k8s spike feature branch.

## QA steps

Regression test IAAS model deployment scenarios.
Regression test v1 k8s charm deployment scenarios
(deploy 2 charms, relate, remove app, destroy model etc)

```console
$ juju add-model t1 microk8s

$ juju deploy ./cockroach --resource cockroachdb-image=cockroachdb/cockroach:v20.1.4

$ kubectl get all,ing,pv,pvc -o wide -n t1
NAME                                READY   STATUS    RESTARTS   AGE   IP            NODE     NOMINATED NODE   READINESS GATES
pod/cockroachdb-0                   2/2     Running   0          19m   10.1.78.110   ubuntu   <none>           <none>
pod/modeloperator-78c8c4b4f-8bgcw   1/1     Running   0          24m   10.1.78.109   ubuntu   <none>           <none>

NAME                            TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)     AGE   SELECTOR
service/cockroachdb             ClusterIP   10.152.183.195   <none>        65535/TCP   19m   juju-app=cockroachdb
service/cockroachdb-endpoints   ClusterIP   None             <none>        <none>      19m   juju-app=cockroachdb
service/modeloperator           ClusterIP   10.152.183.177   <none>        17071/TCP   24m   juju-modeloperator=modeloperator

NAME                            READY   UP-TO-DATE   AVAILABLE   AGE   CONTAINERS      IMAGES                             SELECTOR
deployment.apps/modeloperator   1/1     1            1           24m   juju-operator   ycliuhw/jujud-operator:2.9-beta1   juju-modeloperator=modeloperator

NAME                                      DESIRED   CURRENT   READY   AGE   CONTAINERS      IMAGES                             SELECTOR
replicaset.apps/modeloperator-78c8c4b4f   1         1         1       24m   juju-operator   ycliuhw/jujud-operator:2.9-beta1   juju-modeloperator=modeloperator,pod-template-hash=78c8c4b4f

NAME                           READY   AGE   CONTAINERS                    IMAGES
statefulset.apps/cockroachdb   1/1     19m   juju-unit-agent,cockroachdb   ycliuhw/ubuntu:20.04,cockroachdb/cockroach:v20.1.4

NAME                                                        CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                            STORAGECLASS        REASON   AGE   VOLUMEMODE
persistentvolume/pvc-0797c052-baaa-4ef8-8858-860b38205274   20Gi       RWO            Delete           Bound    controller-k1/storage-controller-0               microk8s-hostpath            31m   Filesystem
persistentvolume/pvc-58ab2fd3-15a9-40eb-8125-3b8478589ca4   1Gi        RWO            Delete           Bound    t1/cockroachdb-database-e94caa18-cockroachdb-0   microk8s-hostpath            19m   Filesystem

NAME                                                                STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS        AGE   VOLUMEMODE
persistentvolumeclaim/cockroachdb-database-e94caa18-cockroachdb-0   Bound    pvc-58ab2fd3-15a9-40eb-8125-3b8478589ca4   1Gi        RWO            microk8s-hostpath   19m   Filesystem

$ kubectl get pod/cockroachdb-0 -n t1 -o json | jq '{containers: .spec.containers, initContainers: .spec.initContainers}'
{
  "containers": [
    {
      "args": [
        "unit",
        "--data-dir",
        "/var/lib/juju"
      ],
      "command": [
        "/usr/bin/k8sagent"
      ],
      "env": [
        {
          "name": "JUJU_CONTAINER_NAMES",
          "value": "cockroachdb"
        }
      ],
      "image": "ycliuhw/ubuntu:20.04",
      "imagePullPolicy": "IfNotPresent",
      "name": "juju-unit-agent",
      "resources": {},
      "terminationMessagePath": "/dev/termination-log",
      "terminationMessagePolicy": "File",
      "volumeMounts": [
        {
          "mountPath": "/usr/bin/k8sagent",
          "name": "juju-data-dir",
          "readOnly": true,
          "subPath": "usr/bin/k8sagent"
        },
        {
          "mountPath": "/var/lib/juju",
          "name": "juju-data-dir",
          "subPath": "var/lib/juju"
        },
        {
          "mountPath": "/var/run/containers",
          "name": "juju-data-dir",
          "subPath": "var/run/containers"
        },
        {
          "mountPath": "/var/lib/juju/storage/database/0",
          "name": "cockroachdb-database-e94caa18"
        }
      ],
      "workingDir": "/var/lib/juju"
    },
    {
      "command": [
        "/usr/bin/pebble"
      ],
      "env": [
        {
          "name": "JUJU_CONTAINER_NAME",
          "value": "cockroachdb"
        }
      ],
      "image": "cockroachdb/cockroach:v20.1.4",
      "imagePullPolicy": "IfNotPresent",
      "name": "cockroachdb",
      "resources": {},
      "terminationMessagePath": "/dev/termination-log",
      "terminationMessagePolicy": "File",
      "volumeMounts": [
        {
          "mountPath": "/usr/bin/pebble",
          "name": "juju-data-dir",
          "readOnly": true,
          "subPath": "usr/bin/pebble"
        },
        {
          "mountPath": "/var/run/container",
          "name": "juju-data-dir",
          "subPath": "var/run/containers/cockroachdb"
        },
        {
          "mountPath": "/cockroach/cockroach-data",
          "name": "cockroachdb-database-e94caa18"
        }
      ]
    }
  ],
  "initContainers": [
    {
      "args": [
        "init"
      ],
      "command": [
        "/opt/k8sagent"
      ],
      "env": [
        {
          "name": "JUJU_CONTAINER_NAMES",
          "value": "cockroachdb"
        },
        {
          "name": "JUJU_K8S_POD_NAME",
          "valueFrom": {
            "fieldRef": {
              "apiVersion": "v1",
              "fieldPath": "metadata.name"
            }
          }
        },
        {
          "name": "JUJU_K8S_POD_UUID",
          "valueFrom": {
            "fieldRef": {
              "apiVersion": "v1",
              "fieldPath": "metadata.uid"
            }
          }
        }
      ],
      "envFrom": [
        {
          "secretRef": {
            "name": "cockroachdb-application-config"
          }
        }
      ],
      "image": "ycliuhw/k8sagent:2.9-beta1",
      "imagePullPolicy": "IfNotPresent",
      "name": "juju-unit-init",
      "resources": {},
      "terminationMessagePath": "/dev/termination-log",
      "terminationMessagePolicy": "File",
      "volumeMounts": [
        {
          "mountPath": "/var/lib/juju",
          "name": "juju-data-dir",
          "subPath": "var/lib/juju"
        },
        {
          "mountPath": "/shared/usr/bin",
          "name": "juju-data-dir",
          "subPath": "usr/bin"
        },
        {
          "mountPath": "/var/run/containers",
          "name": "juju-data-dir",
          "subPath": "var/run/containers"
        }
      ],
      "workingDir": "/var/lib/juju"
    }
  ]
}
```

### CODE REVIEW

**Review by hpidcock — APPROVED** [review]

**Comment by hpidcock** [comment]
$$merge$$

File: apiserver/facades/agent/caasapplication/application.go [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+		}
+		n := fmt.Sprintf("%s/%d", application.Name(), ord)
+		unitName = &n
+	case caas.DeploymentStateless, caas.DeploymentDaemon:
+		// Both handled the same way.
+	default:
+		return errResp(errors.NotSupportedf("unknown deployment type"))
+	}
+
+	var unit Unit = nil
```
Comment (SimonRichardson) [inline apiserver/facades/agent/caasapplication/application.go:122]:
```suggestion
	var unit Unit
```

File: apiserver/facades/agent/caasapplication/application.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+			if addr != "" {
+				addrs = append(addrs, addr)
+			}
+		}
+	}
+
+	caCert, _ := controllerConfig.CACert()
+	version, _ := f.model.AgentVersion()
+	dataDir, _ := paths.DataDir("kubernetes")
+	logDir, _ := paths.LogDir("kubernetes")
```
Comment (SimonRichardson) [inline apiserver/facades/agent/caasapplication/application.go:262]: These don't handle any of the errors...

File: apiserver/facades/agent/caasapplication/application.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+	if err != nil {
+		return errResp(err)
+	}
+
+	apiHostPorts, err := f.state.APIHostPortsForAgents()
+	if err != nil {
+		return errResp(err)
+	}
+
+	addrs := []string(nil)
```
Comment (SimonRichardson) [inline apiserver/facades/agent/caasapplication/application.go:249]:
```suggestion
	var addrs []string
```

File: apiserver/facades/agent/caasapplication/state.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return applicationShim{app}, nil
+}
+
+func (s stateShim) Model() (Model, error) {
+	model, err := s.State.Model()
+	if err != nil {
+		return nil, err
```
Comment (SimonRichardson) [inline apiserver/facades/agent/caasapplication/state.go:67]: Fix all the bare errors being sent back!

```suggestion
		return nil, errors.Trace(err)
```

**Review by SimonRichardson — COMMENTED** [review]
There are a lot more issues here that need resolving here.

File: api/charms/client.go [hunk 5]
```diff
@@ -86,40 +89,6 @@ func (c *Client) ResolveCharms(charms []CharmToResolve) ([]ResolvedCharm, error)
...
-	URL        string
-	Config     *charm.Config
-	Meta       *charm.Meta
-	Actions    *charm.Actions
-	Metrics    *charm.Metrics
-	LXDProfile *charm.LXDProfile
-}
-
-// CharmInfo returns information about the requested charm.
-func (c *Client) CharmInfo(charmURL string) (*CharmInfo, error) {
```
Comment (SimonRichardson) [inline api/charms/client.go:101]: Out of interest, why was this moved?

File: api/common/charms/common.go [hunk 6]
```diff
@@ -239,3 +305,121 @@ func convertCharmLXDProfileDevicesMap(devices map[string]map[string]string) map[
...
+func convertCharmPlatforms(input []string) []charm.Platform {
+	platforms := []charm.Platform(nil)
+	for _, v := range input {
+		platforms = append(platforms, charm.Platform(v))
+	}
+	return platforms
+}
+
+func convertCharmArchitectures(input []string) []charm.Architecture {
+	architectures := []charm.Architecture(nil)
```
Comment (SimonRichardson) [inline api/common/charms/common.go:391]: You know the length of this...

```go
architectures := make([]charm.Architecture, len(input))
```

You should index instead of append...

File: apiserver/common/applicationwatcher.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+// ApplicationFilter to apply to watched model.
+type ApplicationFilter int
+
+const (
+	// ApplicationFilterNone has no filtering.
+	ApplicationFilterNone ApplicationFilter = iota
+	// ApplicationFilterCAASLegacy filters to include only legacy applications.
+	ApplicationFilterCAASLegacy ApplicationFilter = iota
+	// ApplicationFilterCAASEmbedded filters to include only embedded applications.
+	ApplicationFilterCAASEmbedded ApplicationFilter = iota
```
Comment (SimonRichardson) [inline apiserver/common/applicationwatcher.go:27]:
```suggestion
	// ApplicationFilterNone has no filtering.
	ApplicationFilterNone ApplicationFilter = iota
	// ApplicationFilterCAASLegacy filters to include only legacy applications.
	ApplicationFilterCAASLegacy
	// ApplicationFilterCAASEmbedded filters to include only embedded applications.
	ApplicationFilterCAASEmbedded
```
